### PR TITLE
[SPARK-30594][CORE] Do not post SparkListenerBlockUpdated when updateBlockInfo returns false

### PR DESCRIPTION
**Pull request by xuanyuanking**
All credit to @Ngone51, Closes #27306.

### What changes were proposed in this pull request?
If `updateBlockInfo` returns false, which means the `BlockManager` will re-register and report all blocks later. So, we may report two times for the same block, which causes `AppStatusListener` to count used memory for two times, too. As a result, the used memory can exceed the total memory.
So, this PR changes it to not post `SparkListenerBlockUpdated` when `updateBlockInfo` returns false. And, always clean up used memory whenever `AppStatusListener` receives `SparkListenerBlockManagerAdded`.

### Why are the changes needed?
This PR tries to fix negative memory usage in UI (https://user-images.githubusercontent.com/3488126/72131225-95e37e00-33b6-11ea-8708-6e5ed328d1ca.png, see #27144 ). Though, I'm not very sure this is the root cause for #27144 since known information is limited here.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Add new UT.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @cloud-fan @jiangxb1987 @dongjoon-hyun

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 1]
```diff
@@ -98,8 +98,13 @@ class BlockManagerMasterEndpoint(
 
     case _updateBlockInfo @
         UpdateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size) =>
-      context.reply(updateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size))
-      listenerBus.post(SparkListenerBlockUpdated(BlockUpdatedInfo(_updateBlockInfo)))
+      val result = updateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size)
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: nit: maybe `result` -> `isSuccess` ?

**Review by jiangxb1987 — COMMENTED** [review]
LGTM

File: core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala [hunk 2]
```diff
@@ -1657,6 +1657,20 @@ class AppStatusListenerSuite extends SparkFunSuite with BeforeAndAfter {
...
+    val listener = new AppStatusListener(store, conf, true)
+
+    val driver = BlockManagerId(SparkContext.DRIVER_IDENTIFIER, "localhost", 42)
+    listener.onBlockManagerAdded(SparkListenerBlockManagerAdded(
+      time, driver, 42L, Some(43L), Some(44L)))
+
+    check[ExecutorSummaryWrapper](SparkContext.DRIVER_IDENTIFIER) { d =>
+      val memoryMetrics = d.info.memoryMetrics.get
+      assert(memoryMetrics.usedOffHeapStorageMemory == 0)
+      assert(memoryMetrics.usedOnHeapStorageMemory == 0)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala]: Thank you for adding this.
BTW, this fail without `core/src/main/scala/org/apache/spark/status/AppStatusListener.scala`?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Newly added UT should fail without this patch.
Please fix `test("clean up used memory when BlockManager added")`.

**Comment by dongjoon-hyun** [comment]
BTW, @xuanyuanking , @Ngone51 , @jiangxb1987 .
`branch-3.0` cut is `Feature Freeze`, not locking down the bug fixes.
We don't need to hurry for this fix.

cc @gatorsmile

**Comment by HeartSaVioR** [comment]
It's same as my understanding as well. #27306 can just go on.

**Comment by xuanyuanking** [comment]
@dongjoon-hyun Thanks, UT fixed in fecaf20.

Sure, I'll close this one, let's waiting for #27306. cc @cloud-fan

**Pull request closed without merging**